File: app/config/env.py
"""
Environment Configuration
Centralized access to environment variables with type safety and defaults
"""
import os
from typing import Optional
from functools import lru_cache


class Config:
    """Application configuration loaded from environment variables"""
    
    # Service Configuration
    SERVICE_NAME: str = os.getenv("SERVICE_NAME", "cumpair")
    APP_NAME: str = os.getenv("APP_NAME", "Cumpair")
    ENVIRONMENT: str = os.getenv("ENVIRONMENT", "production")
    DEBUG: bool = os.getenv("DEBUG", "false").lower() == "true"
    LOG_LEVEL: str = os.getenv("LOG_LEVEL", "INFO")
    
    # Feature Flags
    DEMO_MODE: bool = os.getenv("DEMO_MODE", "false").lower() == "true"
    
    # Database Configuration
    POSTGRES_DB: str = os.getenv("POSTGRES_DB", "compair")
    POSTGRES_USER: str = os.getenv("POSTGRES_USER", "compair")
    POSTGRES_PASSWORD: str = os.getenv("POSTGRES_PASSWORD", "")
    POSTGRES_HOST: str = os.getenv("POSTGRES_HOST", "localhost")
    POSTGRES_PORT: int = int(os.getenv("POSTGRES_PORT", "5432"))
    
    # Redis Configuration
    REDIS_HOST: str = os.getenv("REDIS_HOST", "localhost")
    REDIS_PORT: int = int(os.getenv("REDIS_PORT", "6379"))
    REDIS_PASSWORD: Optional[str] = os.getenv("REDIS_PASSWORD")
    
    # Web Application
    WEB_PORT: int = int(os.getenv("WEB_PORT", "8000"))
    MAX_FILE_SIZE: int = int(os.getenv("MAX_FILE_SIZE", "10485760"))
    
    # API Configuration
    CORS_ORIGINS: list[str] = os.getenv(
        "CORS_ORIGINS", 
        "http://localhost:3000,http://localhost:8080"
    ).split(",")
    
    # AI Model Configuration
    CLIP_MODEL_NAME: str = os.getenv("CLIP_MODEL_NAME", "ViT-B/32")
    
    @classmethod
    def is_demo_mode(cls) -> bool:
        """Check if application is running in demo mode"""
        return cls.DEMO_MODE
    
    @classmethod
    def get_service_name(cls) -> str:
        """Get the service name for metrics and logging"""
        return cls.SERVICE_NAME
    
    @classmethod
    def get_database_url(cls) -> str:
        """Get the database connection URL"""
        password = f":{cls.POSTGRES_PASSWORD}" if cls.POSTGRES_PASSWORD else ""
        return (
            f"postgresql://{cls.POSTGRES_USER}{password}"
            f"@{cls.POSTGRES_HOST}:{cls.POSTGRES_PORT}/{cls.POSTGRES_DB}"
        )
    
    @classmethod
    def get_redis_url(cls) -> str:
        """Get the Redis connection URL"""
        password = f":{cls.REDIS_PASSWORD}@" if cls.REDIS_PASSWORD else ""
        return f"redis://{password}{cls.REDIS_HOST}:{cls.REDIS_PORT}/0"


@lru_cache()
def get_config() -> Config:
    """Get cached configuration instance"""
    return Config()


# Export convenience functions
def is_demo_mode() -> bool:
    """Check if running in demo mode"""
    return get_config().is_demo_mode()


def get_service_name() -> str:
    """Get the service name"""
    return get_config().get_service_name()


# Log configuration on import (only in debug mode)
if Config.DEBUG:
    print(f"[Config] Environment configuration loaded:")
    print(f"  - Service Name: {Config.SERVICE_NAME}")
    print(f"  - Demo Mode: {Config.DEMO_MODE}")
    print(f"  - Environment: {Config.ENVIRONMENT}")
    print(f"  - Debug: {Config.DEBUG}")

File: frontend/src/config/env.ts
/**
 * Environment Configuration
 * Centralized access to environment variables with type safety and defaults
 */

export const config = {
  // Service Configuration
  serviceName: import.meta.env.VITE_SERVICE_NAME || 'cumpair',
  demoMode: import.meta.env.VITE_DEMO_MODE === 'true',
  
  // API Configuration
  apiUrl: import.meta.env.VITE_API_URL || 'http://localhost:8000',
  apiTimeout: parseInt(import.meta.env.VITE_API_TIMEOUT || '30000', 10),
  
  // Feature Flags
  features: {
    realTimeSearch: import.meta.env.VITE_ENABLE_REAL_TIME_SEARCH === 'true',
    priceAlerts: import.meta.env.VITE_ENABLE_PRICE_ALERTS === 'true',
    analytics: import.meta.env.VITE_ENABLE_ANALYTICS === 'true',
    geolocation: import.meta.env.VITE_ENABLE_GEOLOCATION === 'true',
    demoOldComponents: import.meta.env.VITE_DEMO_OLD_COMPONENTS === 'true',
  },
  
  // Search Configuration
  search: {
    defaultSites: (import.meta.env.VITE_DEFAULT_SEARCH_SITES || 'amazon,walmart,ebay').split(','),
    minSimilarityScore: parseFloat(import.meta.env.VITE_MIN_SIMILARITY_SCORE || '0.7'),
    maxResults: parseInt(import.meta.env.VITE_MAX_SEARCH_RESULTS || '20', 10),
  },
  
  // Upload Configuration
  upload: {
    maxFileSize: parseInt(import.meta.env.VITE_MAX_FILE_SIZE || '10485760', 10),
    allowedTypes: (import.meta.env.VITE_ALLOWED_FILE_TYPES || 'image/jpeg,image/png,image/webp,image/gif').split(','),
  },
  
  // Development Settings
  debug: import.meta.env.VITE_DEBUG_MODE === 'true',
  logLevel: import.meta.env.VITE_LOG_LEVEL || 'info',
} as const;

// Helper function to check if running in demo mode
export const isDemoMode = () => config.demoMode;

// Helper function to get service name
export const getServiceName = () => config.serviceName;

// Helper function to check feature flags
export const isFeatureEnabled = (feature: keyof typeof config.features) => {
  return config.features[feature];
};

// Export convenience constants
export const API_BASE_URL = config.apiUrl;
export const WS_BASE_URL = config.apiUrl.replace('http', 'ws');

// Log configuration on load (only in debug mode)
if (config.debug) {
  console.log('[Config] Environment configuration loaded:', {
    serviceName: config.serviceName,
    demoMode: config.demoMode,
    apiUrl: config.apiUrl,
    features: config.features,
  });
}
